stdlib/public/Concurrency/TaskPrivate.h

```diff
@@ -168,35 +169,96 @@ class TaskFutureWaitAsyncContext : public AsyncContext {
+  ///
+  ///
+  /// 64 bit systems with SWIFT_CONCURRENCY_ENABLE_PRIORITY_ESCALATION=1
+  ///
+  ///         Flags               Drain Lock          TaskStatusRecord *
+  /// |----------------------|-------------------|----------------------|
+  ///          32 bits                32 bits              64 bits
+  ///
+  /// 32 bit systems with SWIFT_CONCURRENCY_ENABLE_PRIORITY_ESCALATION=0
+  ///
+  ///          Flags             TaskStatusRecord *
+  /// |----------------------|----------------------|
+  ///          32 bits                32 bits
+  //
+  /// 64 bit systems with SWIFT_CONCURRENCY_ENABLE_PRIORITY_ESCALATION=0
+  ///
+  ///         Flags                  Unused            TaskStatusRecord *
+  /// |----------------------|----------------------|---------------------|
+  ///         32 bits                 32 bits                64 bits
+  ///
+  /// Note: This layout is intentional since we can then do 64bit atomics on
+  /// just the top 64 bits of atomic state when
+  /// SWIFT_CONCURRENCY_ENABLE_PRIORITY_ESCALATION=1 if we don't need to
+  /// coordinate with the status record pointer.
+
+  enum : uint32_t {
     /// The max priority of the task. This is always >= basePriority in the task
     PriorityMask = 0xFF,
 
     /// Has the task been cancelled?
     IsCancelled = 0x100,
 
     /// Whether the task status is "locked", meaning that further
-    /// accesses need to wait on the task status record lock
-    IsLocked = 0x200,
+    /// accesses need to wait on the task status record lock. This is separate
+    /// from the drain lock of the task.
+    IsStatusRecordLocked = 0x200,
 
-    /// Whether the running priority has been escalated above the
-    /// priority recorded in the Job header.
+    /// Whether the running priority has been escalated above the previous max
+    /// priority in the task status
     IsEscalated = 0x400,
 
-    /// Whether the task is actively running.
-    /// We don't really need to be tracking this in the runtime right
-    /// now, but we will need to eventually track enough information to
-    /// escalate the thread that's running a task, so doing the stores
-    /// necessary to maintain this gives us a more realistic baseline
-    /// for performance.
+#if !SWIFT_CONCURRENCY_ENABLE_PRIORITY_ESCALATION
+    /// Whether the task is actively running. On systems which cannot track the
+    /// identity of the drainer (see above), we use one bit in the flags to track
+    /// whether or not task is running. Otherwise, the drain lock tells us
+    /// whether or not it is running.
     IsRunning = 0x800,
+#endif
+
+    /// Task is enqueued somewhere - either in the cooperative pool, or in an
+    /// actor. This bit is only clear when a task stops running and becomes
+    /// suspended, or when it is completed.
+    ///
+    /// Basically, "enqueued" means that the task is runnable or running. Using
+    /// same terminology and semantics as logic in dispatch
```

Comment:
Yeah, this kind of "it's just like Dispatch" comment isn't really okay here, because other maintainers of this code (like me!) don't have any idea what it means.

In general, I would strongly prefer that we not widely introduce Dispatch jargon like "drainer" into the Swift runtime except when specifically necessary to talk about an interaction with Dispatch.  Nothing "drains" a task.

I'm not sure I understand the value of tracking whether we're currently enqueued given that we're not tracking the identity of the queue (yet).

Reply:
Yea I can work on better names and comments. I just started with what felt natural to me, in order to things bootstrapped and working end-to-end.

> I'm not sure I understand the value of tracking whether we're currently enqueued given that we're not tracking the identity of the queue (yet).

a) I think It is helpful for us to understand whether the task is in the state of suspended, runnable or running and therefore what kinds of action we need to take for priority escalation. We currently don't do anything for the `runnable` case (ie no stealing) but it is helpful bookkeeping to start having.

b) It also helps us identity whether the Job fields should make sense or not since they really only are of value in enqueued jobs when inspecting the structure from outside.